contracts/universalResolver/ERC3668Caller.sol

```diff
@@ -0,0 +1,262 @@
+                errorData,
+                internalExtraData
+            );
+        }
+
+        // unknown error, propagate the revert
+        LowLevelCallUtils.propagateRevert(errorData);
+    }
+
+    /// @notice The ERC3668 callback function.
+    /// @param response The response from the ERC3668 call, arbitrary data.
+    /// @param extraData The passthrough data from `call`, encoded as `(address target, uint256 callbackFunctions, bytes internalExtraData, bytes externalExtraData)`.
+    function callback(
+        bytes calldata response,
+        bytes calldata extraData
+    ) external view {
+        (
+            address target,
+            uint256 callbackFunctions,
+            bytes memory internalExtraData,
+            bytes memory externalExtraData
+        ) = _getExtraData(extraData);
+
+        bytes4 validateResponseFunction = bytes4(
+            uint32(
+                (callbackFunctions >> VALIDATE_RESPONSE_FUNCTION_OFFSET) &
+                    FUNCTIONS_MASK
+            )
+        );
+        if (validateResponseFunction != bytes4(0))
+            _validateResponse(validateResponseFunction, response);
+
+        bytes4 externalCallbackFunction = bytes4(
+            uint32(
+                (callbackFunctions >> EXTERNAL_CALLBACK_FUNCTION_OFFSET) &
+                    FUNCTIONS_MASK
+            )
+        );
+        return
+            call(
+                target,
+                0,
+                bytes.concat(
+                    externalCallbackFunction,
+                    abi.encode(response, externalExtraData)
+                ),
+                internalExtraData,
+                uint128(callbackFunctions)
+            );
+    }
+
+    /// @notice Defines the callback functions to be used with `call`.
+    /// If only using `internalCallbackFunction`, use `uint32(internalCallbackFunction)` instead of this function.
+    /// @param internalCallbackFunction The internal callback function.
+    /// @param calldataRewriteFunction The calldata rewrite function.
+    /// @param failureCallbackFunction The failure callback function.
+    /// @param validateResponseFunction The validate response function.
+    /// @return The encoded callback functions.
+    function createUserCallbackFunctions(
+        bytes4 internalCallbackFunction,
+        bytes4 calldataRewriteFunction,
+        bytes4 failureCallbackFunction,
+        bytes4 validateResponseFunction
+    ) internal pure returns (uint128) {
+        return
+            (uint128(uint32(validateResponseFunction)) <<
+                VALIDATE_RESPONSE_FUNCTION_OFFSET) |
+            (uint128(uint32(failureCallbackFunction)) <<
+                FAILURE_CALLBACK_FUNCTION_OFFSET) |
+            (uint128(uint32(calldataRewriteFunction)) <<
+                CALLDATA_REWRITE_FUNCTION_OFFSET) |
+            uint128(uint32(internalCallbackFunction));
+    }
+
+    /// @dev Formats the calldata for the offchain lookup.
+    function _formatCalldata(
```

Comment:
do you just mean the execution of the code for it should be inlined? or are you talking about changing the api

Reply:
It's set up to do a staticcall to this same contract. Why? Why not just call the target function directly?

Reply:
the contract is meant to be general purpose, and it makes for a cleaner api if the format/validate callbacks are passed as parameters. since the UR only uses both of those once though, i suppose they could just be overridable functions